RENO-1881: Update Pagination component (in progress)

misselbeck opened the pull request:
Fixes #437 

## **This PR does the following:**

I took a first pass (rough draft) at this based on the Figma design. There seems to be some inferred logic that the component would need to handle, such as navigating between items, applying an active state, and passing back the selected item to the consuming application to apply to its own app.

I came up with a component that takes in a page count and a callback function as props.

### Props

| Name  | Type | Description
| ------------- | ------------- |------------- |
| `pageCount`  | `Number`  |  The total number of items. |
| `onPageChange` | `Function`  | The method to call when an item is selected.<br>Exposes the selected page as an argument.

### Example Usage
```
<DS.Pagination
  pageCount={10}
  onPageChange={pageUpdate}
/>
```

### Result

![nypl-pagination](https://user-images.githubusercontent.com/1325185/100005946-c0081980-2d97-11eb-86dc-ddda33b0bac2.gif)

### To-Do

**Truncation**

Based on the Figma designs, there is also some logic needed to truncate long lists, using ellipses as placeholders, and showing just the first, last, active page + 2 siblings at a time on each side (?).

<img width="227" alt="figma" src="https://user-images.githubusercontent.com/1325185/100006705-ea0e0b80-2d98-11eb-82f2-3599ba62cdd8.png">

This looks like it changes a bit if you are at the beginning or the end of the list, in which it only shows one set of ellipses (either just after the first item, or just before the last item).

I couldn't quite work out the logic in the first pass. After a little investigation, it seems like maybe we want something similar to how the Pagination works in the material-ui project.

Does the logic here look like it would meet the Figma requirements?
https://material-ui.com/components/pagination/#pagination-ranges

I can try to implement something similar in this component (though it looks tricky).

**Previous / Next**

Based on the Figma designs, the Previous button should not display if at the first item, and the Next button should not display if at the last item.

I have this working, but adding/removing the buttons makes the pager "jump." As a page selection most likely will not trigger a page refresh in a React app, maybe it would be better to just have the Previous/Next buttons just become disabled as needed, rather than removing completely?

### Note

I have not touched the Storybook/Tests for this component yet. They are outdated in this branch at the moment.

[Pagination in Storybook](https://deploy-preview-442--stoic-murdock-c7f044.netlify.app/?path=/story/pagination--pagination)

misselbeck committed:
RENO-1881: Update Pagination component (in progress)

hokei commented on src/components/Pagination/Pagination.tsx:
```diff
@@ -1,42 +1,142 @@
+
+        return (
+            <li key={index + 1}>
+                <Button
```
I think you want to be using `<Link>` here instead of `<Button>`, because the underlying native element you want is an `<a>` element rather than a `<button>` element.  (Same goes for "next" and "previous")

misselbeck replied:
I made this change. All the page items are now using the `<Link>` component.

hokei commented on src/components/Pagination/Pagination.tsx:
```diff
@@ -1,42 +1,142 @@
+                    [this.props.className]
+                )}
+            >
+                <ul>
```
~~This is kind of a nit because I'm not sure it matters, ultimately, but `<ul>` doesn't seem semantically correct here.  You can't re-order the `<li>` items at will.  I'm seeing that Amazon is using `ul` anyway, while Google is using `table`, and both are weird to me.  I think I like 
https://react.semantic-ui.com/addons/pagination/
the best, where it's just a div with a bunch of links in it.  This is not tabular data, so table seems weird.  `ol` might be fine? Thoughts?~~

edit: I'm a dumb person who got ol and ul mixed up cuz clearly I can't spell.  `<ul>` is clearly fine here.  Sorry, ignore this.

misselbeck committed:
Update component to use <Link> instead of button

misselbeck committed:
Add an initial story for review

misselbeck committed:
Remove test for now so build runs

misselbeck commented:
**To Do**
- [x] Add a Storybook entry
- [x] Add truncation logic and display
- [x] Write a test file

helenvholmes commented:
@misselbeck, I've updated your PR description to add the multidev preview. You can grab that link in the future by scrolling in the checks area to `netlify/stoic-murdock...etc` and hitting the 'Details' link.

helenvholmes commented:
> Does the logic here look like it would meet the Figma requirements?
> https://material-ui.com/components/pagination/#pagination-ranges
> 
> I can try to implement something similar in this component (though it looks tricky).

We styled our designs off of Google's pagination (since they're so good at handling large search result sets like this), so yes! I would go even further and say that if our designs are _missing_ something, this is the component we'll be looking to for guidance. Best of luck, I think we're all aware this component is going to be pretty tricky. ☹️
 
> **Previous / Next**
> 
> Based on the Figma designs, the Previous button should not display if at the first item, and the Next button should not display if at the last item.
> 
> I have this working, but adding/removing the buttons makes the pager "jump." As a page selection most likely will not trigger a page refresh in a React app, maybe it would be better to just have the Previous/Next buttons just become disabled as needed, rather than removing completely?

That's a good suggestion, so I've created [UX-162](https://jira.nypl.org/browse/UX-162) to make this update.

misselbeck committed:
Merge branch 'development' of github.com:NYPL/nypl-design-system into RENO-1881-update-pagination-component

helenvholmes commented:
An odd thing happening on the multidev is that the knobs for this component are appearing across of the other components on the multidev... super weird

misselbeck committed:
RENO-1881: Add truncation logic to Pagination component

misselbeck committed:
RENO-1881: Update default pageCount to 100

hokei commented on src/components/Pagination/Pagination.stories.tsx:
```diff
@@ -1,99 +1,29 @@
-const PaginationTemplate = ({ children, ...args }) => (
-    <Pagination {...args}>{children}</Pagination>
+const PaginationTemplate: Story<PaginationProps> = (args) => (
+    <Pagination {...args}></Pagination>
```
Is it a Storybook requirement to pass the props in with spread syntax?  If not, I would prefer to pass the args in explicitly so that anyone looking at this documentation knows what the valid and expected args are!

misselbeck replied:
It looks like the other components using a Storybook template are doing it this way. I can investigate if there is another approach.

misselbeck replied:
Args now passed in specifically, as requested.

misselbeck commented:
> An odd thing happening on the multidev is that the knobs for this component are appearing across of the other components on the multidev... super weird

I can't replicate this. Let me know if this is ongoing. I can revisit my story file for any anomalies.

misselbeck commented:
Added the truncation logic.

To Do:

- [x] Write a test file
- [x] Get linting to work!

I noticed that the `.prettierrc` settings are adding semicolons to my code (as well as 4 tab indentation, etc). But then when running `npm run lint`, it complains about the semicolons. This also fails during the CI/Lint (push).

Just me? Any tricks to getting this setup properly with VS Code?

nypl-wluisi commented:
@misselbeck Whenever you get a chance, can you also create a PR off the location finder repo that uses this branch and updates the pagination on location finder? You see how I did similar here: https://github.com/NYPL/dxp-react-search/pull/81/files

In your `package.json` you can point the DS to use this github branch, so something like:
`"@nypl/design-system-react-components": "git+https://git@github.com:NYPL/nypl-design-system.git#RENO-1881-update-pagination-component"`

And you'd have to update this component:
https://github.com/NYPL/dxp-react-search/blob/development/src/components/location-finder/Locations/LocationsPagination.js

We can probs track the work in another ticket, but just doing a basic sanity check that this component will work in LF would be helpful.

helenvholmes commented:
> I noticed that the `.prettierrc` settings are adding semicolons to my code (as well as 4 tab indentation, etc). But then when running `npm run lint`, it complains about the semicolons. This also fails during the CI/Lint (push).
> 
> Just me? Any tricks to getting this setup properly with VS Code?

Grrr, this is mysterious. You might need to crowd-source solutions in #designs-system, I've never seen something like this :/

helenvholmes commented:
> > I noticed that the `.prettierrc` settings are adding semicolons to my code (as well as 4 tab indentation, etc). But then when running `npm run lint`, it complains about the semicolons. This also fails during the CI/Lint (push).
> > Just me? Any tricks to getting this setup properly with VS Code?
> 
> Grrr, this is mysterious. You might need to crowd-source solutions in #designs-system, I've never seen something like this :/

Did some research and stumbled across [this thread](https://github.com/palantir/tslint/issues/2431). I think we might need to update our `tslint.json` L14 to `"semicolon": [true, "always", "strict-bound-class-methods"],`. @misselbeck, wanna make that change and see if it fixes `npm test` for you?

misselbeck commented:
> > > I noticed that the `.prettierrc` settings are adding semicolons to my code (as well as 4 tab indentation, etc). But then when running `npm run lint`, it complains about the semicolons. This also fails during the CI/Lint (push).
> > > Just me? Any tricks to getting this setup properly with VS Code?
> > 
> > 
> > Grrr, this is mysterious. You might need to crowd-source solutions in #designs-system, I've never seen something like this :/
> 
> Did some research and stumbled across [this thread](https://github.com/palantir/tslint/issues/2431). I think we might need to update our `tslint.json` L14 to `"semicolon": [true, "always", "strict-bound-class-methods"],`. @misselbeck, wanna make that change and see if it fixes `npm test` for you?

This looks like it fixes the issue. Thanks for tracking that down! Should I commit it? Here or in another branch?

helenvholmes commented:
> This looks like it fixes the issue. Thanks for tracking that down! Should I commit it? Here or in another branch?

I say ship it! 🚀 let's do it here, why not.

misselbeck commented:
> @misselbeck Whenever you get a chance, can you also create a PR off the location finder repo that uses this branch and updates the pagination on location finder? You see how I did similar here: https://github.com/NYPL/dxp-react-search/pull/81/files
> 
> In your `package.json` you can point the DS to use this github branch, so something like:
> `"@nypl/design-system-react-components": "git+https://git@github.com:NYPL/nypl-design-system.git#RENO-1881-update-pagination-component"`
> 
> And you'd have to update this component:
> https://github.com/NYPL/dxp-react-search/blob/development/src/components/location-finder/Locations/LocationsPagination.js
> 
> We can probs track the work in another ticket, but just doing a basic sanity check that this component will work in LF would be helpful.

This is a good idea. I'll start on that.

helenvholmes commented:
@misselbeck based on our conversation in this thread I've updated the `Pagination` designs to not shift the next/previous buttons and adhere more closely to [Google Material Design's component](https://material-ui.com/components/pagination/). The designs for that are [here](https://www.figma.com/file/qShodlfNCJHb8n03IFyApM/Main?node-id=17226%3A932).

misselbeck committed:
RENO-1881: Re-work Pagination component to not rely on state

misselbeck committed:
RENO-1881: Update Pagination story to use new prop; pass args in explicitly rather than spread syntax

misselbeck committed:
RENO-1881: Fix for semicolon complaint from tslint

misselbeck committed:
RENO-1881: Remove unnecessary commented code

misselbeck committed:
RENO-1881: Add Pagination component tests

misselbeck committed:
Merge branch 'development' of github.com:NYPL/nypl-design-system into RENO-1881-update-pagination-component

misselbeck committed:
RENO-1881: Lint adjustments to scss

misselbeck committed:
RENO-1881: Remove default list styling

misselbeck committed:
RENO-1881: Storybook minor aesthetics

misselbeck commented:
After [wiring up the Pagination in the Location Finder app](https://github.com/NYPL/dxp-react-search/pull/84), I think it's best to go with what @hokei suggested on Slack, and not rely on state and let the consuming app pass the currentPage. All the previous functionality works just as well doing this way, and it avoids potential  issues with keeping the state in sync after re-renders, etc.

- Re-worked the component to remove state and added a new prop for the currentPage
- Updated the Pagination Story
- Wrote some tests

misselbeck committed:
RENO-1881: Update pagination disabled item color to match Figma spec

misselbeck committed:
RENO-1881: CHANGELOG entry

misselbeck committed:
Merge branch 'development' into RENO-1881-update-pagination-component

helenvholmes commented on src/components/Pagination/_Pagination.scss:
```diff
@@ -11,13 +11,32 @@
 
-    .label {
-        @include space-stack-none;
+    ul {
```
Can we give these block modifiers instead of styling the elements themselves? 

I might call this `&__numbers`, for example.

Same comments for L19 + L22.

misselbeck replied:
I updated the component and styles to make better use of the BEM function.

helenvholmes commented on src/components/Pagination/Pagination.tsx:
```diff
@@ -1,42 +1,209 @@
+                <Link
+                    attributes={{ ...pageAttributes }}
+                    className={currentPage === item ? "selected" : null}
+                    href="#"
```
Does this being specified here prevent users from updating the route in the url bar?

misselbeck replied:
It doesn't affect the url bar. There is an event handler on the click event that prevents that.

helenvholmes commented on src/components/Pagination/Pagination.stories.tsx:
```diff
@@ -1,99 +1,30 @@
-        </Button>
-    ),
-    children: selectOptions,
+    pageCount: 100,
```
For whatever reason I'm not able to click around to different pages on Storybook.

misselbeck replied:
I recently re-worked the component to remove having its own state, and instead added a new prop for the currentPage. There are some further details in the comment a little further above, but basically, after testing the Pagination component in the Location Finder, it seemed better to let the consuming application keep track.

Because the component no longer has its own state, you have to use the Storybook control for currentPage to simulate navigating through the Pagination.

I could look into adding a useState hook into the Storybook file so the example would be clickable. Should I give that a shot?

helenvholmes replied:
> I recently re-worked the component to remove having its own state, and instead added a new prop for the currentPage. There are some further details in the comment a little further above, but basically, after testing the Pagination component in the Location Finder, it seemed better to let the consuming application keep track.
On the same page here—this is definitely the right direction for this component.

> Because the component no longer has its own state, you have to use the Storybook control for currentPage to simulate navigating through the Pagination.
> 
> I could look into adding a useState hook into the Storybook file so the example would be clickable. Should I give that a shot?
I would really appreciate you giving this a shot—I know that I wanted to understand how to click around in the component and see it in action. I suspect I won't be the only one confused.

misselbeck replied:
I was able to add the hooks to the Story, so that the Pagination component can be clicked in the example and it will update accordingly.

helenvholmes reviewed (changes requested):
Some notes on the SCSS but mostly questions.

misselbeck committed:
RENO-1881: Update Pagination list elements to use BEM naming convention

misselbeck committed:
Merge branch 'development' of github.com:NYPL/nypl-design-system into RENO-1881-update-pagination-component

misselbeck committed:
Adds a couple hooks to the component Storybook example for better interaction

misselbeck committed:
Merge branch 'development' of github.com:NYPL/nypl-design-system into RENO-1881-update-pagination-component

misselbeck committed:
Merge branch 'development' of github.com:NYPL/nypl-design-system into RENO-1881-update-pagination-component

# Conflicts:
#	CHANGELOG.md

misselbeck commented:
Let me know if there are any further changes needed.

helenvholmes commented on src/components/Pagination/_Pagination.scss:
```diff
@@ -11,13 +11,32 @@
+            color: var(--ui-black);
+        }
+
+        &[role="button"] {
```
Perhaps this should be moved into `_Link.scss`, since it seems like something you'd want to happen universally.

helenvholmes commented on src/components/StatusBadge/StatusBadge.test.tsx:
```diff
@@ -10,15 +10,16 @@ describe("StatusBadge Test", () => {
     it("Shows status badge", () => {
         wrapper = Enzyme.shallow(
-            <StatusBadge level={"low"} statusBadgeText={"Registration Required"} />
+            <StatusBadge
```
Are these `StatusBadge` changes something you've done, @misselbeck, or something showing up from a funky merge?...

misselbeck replied:
I didn't work on the `StatusBadge `component at all. It looks like these files were included as part of a merge from `development` that had a conflict with the CHANGELOG.

misselbeck replied:
Prettier formatting differences, perhaps?

helenvholmes replied:
Ahh, got it.

helenvholmes reviewed (approved):
👏🏻

This looks incredible. Thank you Matt. I left a couple of comments—one about where some CSS should live and some questions about some changes to `StatusBadge`, but both are pretty minor.

misselbeck committed:
RENO-1881: Move link button cursor style to Link component for global application
